File: 16. Binary Subarray with Sum = Goal.cpp
// Topic: Number of subarrays with sum = goal (binary array)
#include <bits/stdc++.h>
using namespace std;

int numSubarraysWithSum(vector<int>& A, int S)
{
    unordered_map<int, int> cnt;
    cnt[0] = 1;
    int prefix = 0, ans = 0;
    for (int x : A)
    {
        prefix += x;
        if (cnt.count(prefix - S))
            ans += cnt[prefix - S];
        cnt[prefix]++;
    }
    return ans;
}

int main()
{
    vector<int> A = {1, 0, 1, 0, 1};
    int S = 2;
    cout << "Count: " << numSubarraysWithSum(A, S) << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: 16. Binary Subarray with Sum = Goal.py
# Topic: Number of subarrays with sum = goal (binary array)

from collections import defaultdict

def num_subarrays_with_sum(A, S):
    cnt = defaultdict(int)
    cnt[0] = 1
    prefix = 0
    ans = 0
    for x in A:
        prefix += x
        ans += cnt.get(prefix - S, 0)
        cnt[prefix] += 1
    return ans

if __name__ == '__main__':
    A = [1,0,1,0,1]
    S = 2
    print('Count:', num_subarrays_with_sum(A,S))

# Time Complexity: O(n)
# Space Complexity: O(n)
